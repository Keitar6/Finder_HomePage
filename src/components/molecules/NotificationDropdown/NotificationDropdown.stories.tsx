import { Meta, StoryObj } from '@storybook/react';
import GlobalStyle from '../../../theme/globalStyles';
import ThemeProvider from '../../../theme/ThemeContext';
import NotificationDropdown from './NotificationDropdown';

const meta = {
  title: 'Molecules/NotificationDropdown',
  component: NotificationDropdown,
  tags: ['autodocs'],
  args: { placeholder: 'Search flats...' },
  argTypes: {

  },
  decorators: [
    (Story) => (
      <ThemeProvider>
        <GlobalStyle />
        {Story()}
      </ThemeProvider>
    ),
  ],
  parameters: {
    backgrounds: { default: 'light' },
  },
} satisfies Meta<typeof NotificationDropdown>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};
