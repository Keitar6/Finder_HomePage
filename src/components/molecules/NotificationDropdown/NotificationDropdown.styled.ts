import styled from 'styled-components';
import initialTheme from '../../../theme/initialTheme';

export const DropdownBox = styled.div`
  display: block;
  height: 100%;
  position: relative;
  height: auto;
  max-height: 15.625;
  max-width: 25rem;
`;

export const Wrapper = styled.div`
  position: absolute;
  top: 3.4375rem;
  right: -9px;
  min-width: 14rem;
  box-shadow: 0 5px 10px 0 ${initialTheme.palette.iconInputGrey};
  border-radius: 10px;
`;

export const Container = styled.div`
  background-color: ${initialTheme.palette.light};
  border: 1px solid ${initialTheme.palette.inputGrey};
  border-radius: 6px;

  div:first-child {
    border-radius: 6px 6px 0 0;
  }
  div:last-child {
    border-radius: 0 0 6px 6px;
    border-bottom: transparent;
  }
`;

export const Bell = styled.div`
  z-index: 10;
`;
