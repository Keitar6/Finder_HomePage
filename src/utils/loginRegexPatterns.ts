import { InputVariant, RegexPat } from '../components/atoms/Input/Input.types';
import { Getters } from '../types/shared.types';

type LoginRegexProps = Getters<InputVariant, RegexPat>;

const loginRegexPattern: LoginRegexProps = {
  search: {
    pattern: /^[a-zA-Z0-9]+$/,
    required: true,
    maxLength: 36,
    minLength: 8,
  },
};

export default loginRegexPattern;
