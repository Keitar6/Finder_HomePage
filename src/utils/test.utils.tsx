import { PropsWithChildren } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import ThemeProvider from '../theme/ThemeContext';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialUrl?: string;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    initialUrl = '/',
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  const Wrapper = ({ children }: PropsWithChildren<object>): JSX.Element => (
    <ThemeProvider>
        {children}
    </ThemeProvider>
  );

  return { ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';
