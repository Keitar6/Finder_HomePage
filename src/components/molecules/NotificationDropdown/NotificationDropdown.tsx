import { animations } from '../../../theme/animations';
import Icon from '../../atoms/Icon/Icon';
import NotificationCard from '../NotificationCard/NotificationCard';
import * as Styled from './NotificationDropdown.styled';
import { FC, useContext, useEffect, useRef, useState } from 'react';
import { Badge, IconButton } from '@mui/material';
import { GeneralContext } from '../../../contexts/GeneralContext';

const NotificationDropdown: FC = () => {
  const [isDropdownOpen, setIsDropdownOpen] = useState<boolean>(false);
  const [notificationsHasBeenSeen, setNotificationsHasBeenSeen] = useState<boolean>(false);
  const { notifications } = useContext(GeneralContext);

  const ref = useRef<HTMLDivElement | null>(null);
  const onClickOutside = () => {
    setIsDropdownOpen(false);
  };

  const changeStateOfNotificationDropdown = () => {
    setIsDropdownOpen(!isDropdownOpen);
    setNotificationsHasBeenSeen(true);
  };

  const notificationsLabel = (count: number) => {
    if (count === 0) {
      return 'no notifications';
    }
    if (count > 99) {
      return 'more than 99 notifications';
    }
    return `${count} notifications`;
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent): void => {
      if (ref.current && !ref.current.contains(event.target as Node)) {
        onClickOutside();
      }
    };
    document.addEventListener('click', handleClickOutside, true);
    return () => {
      document.removeEventListener('click', handleClickOutside, true);
    };
  }, []);

  return (
    <Styled.DropdownBox ref={ref}>
      <Styled.Bell>
        <IconButton
          aria-label={notificationsLabel(100)}
          onClick={changeStateOfNotificationDropdown}
        >
          <Badge
            badgeContent={!notificationsHasBeenSeen ? notifications.length : 0}
            color='error'
            sx={{ animation: animations.pulse }}
          >
            <Icon color='text' name='bell' />
          </Badge>
        </IconButton>
      </Styled.Bell>
      {isDropdownOpen ? (
        <Styled.Wrapper>
          <Styled.Container>
            {notifications.map((notification, index) => (
              <NotificationCard
                key={`Notification:${index} - ${notification.title}`}
                data={notification}
              />
            ))}
          </Styled.Container>
        </Styled.Wrapper>
      ) : null}
    </Styled.DropdownBox>
  );
};
export default NotificationDropdown;
