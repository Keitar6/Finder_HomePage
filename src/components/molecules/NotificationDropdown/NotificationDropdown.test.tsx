import { cleanup } from '@testing-library/react';
import { renderWithProviders } from '../../../utils/test.utils';
import NotificationDropdown from './NotificationDropdown';

describe('NotificationDropdown', () => {
  afterEach(cleanup);
  describe('NotificationDropdown component', () => {

    const NotificationDropdownRender: any = () => renderWithProviders(<NotificationDropdown />, {});
    it('should render correctly', () => {
      const NotificationDropdown = NotificationDropdownRender();
      expect(NotificationDropdown).toMatchSnapshot();
    });
  });
});
