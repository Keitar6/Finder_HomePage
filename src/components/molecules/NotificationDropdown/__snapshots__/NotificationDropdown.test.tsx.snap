// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationDropdown NotificationDropdown component should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-kAyceB hyVyvf"
      >
        <div
          class="sc-dhKdcB eLqhmL"
        >
          <button
            aria-label="more than 99 notifications"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiBadge-root css-ui7n8a-MuiBadge-root"
            >
              <span />
              <span
                class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorError css-to4x1n-MuiBadge-badge"
              >
                2
              </span>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-kAyceB hyVyvf"
    >
      <div
        class="sc-dhKdcB eLqhmL"
      >
        <button
          aria-label="more than 99 notifications"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiBadge-root css-ui7n8a-MuiBadge-root"
          >
            <span />
            <span
              class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorError css-to4x1n-MuiBadge-badge"
            >
              2
            </span>
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
